src/main/java/io/strimzi/kafka/quotas/VolumeSource.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.kafka.quotas;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.DescribeClusterResult;
+import org.apache.kafka.clients.admin.DescribeLogDirsResult;
+import org.apache.kafka.clients.admin.LogDirDescription;
+import org.apache.kafka.common.Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.stream.Collectors.toSet;
+
+
+/**
+ * Leverages <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827</a>
+ * to gather volume usage statistics for each Kafka log dir reported by the cluster.
+ * <p>
+ * A listener is registered with this volume source to act on the disk usage information.
+ */
+public class VolumeSource implements Runnable {
+
+    private final VolumeObserver volumeObserver;
+    private final Admin admin;
+    private final int timeout;
+    private final TimeUnit timeoutUnit;
+
+    private static final Logger log = LoggerFactory.getLogger(VolumeSource.class);
+
+    /**
+     * @param admin          The Kafka Admin client to be used for gathering information.
+     * @param volumeObserver the listener to be notified of the volume usage
+     * @param timeout        how long should we wait for cluster information
+     * @param timeoutUnit    What unit is the timeout configured in
+     */
+    @SuppressFBWarnings("EI_EXPOSE_REP2") //Injecting the dependency is the right move as it can be shared
+    public VolumeSource(Admin admin, VolumeObserver volumeObserver, int timeout, TimeUnit timeoutUnit) {
+        this.volumeObserver = volumeObserver;
+        this.admin = admin;
+        this.timeout = timeout;
+        this.timeoutUnit = timeoutUnit;
+    }
+
+    @Override
+    public void run() {
+        log.debug("Attempting to describe cluster");
+        final DescribeClusterResult clusterResult = admin.describeCluster();
+        try {
+            clusterResult.nodes().whenComplete((nodes, throwable) -> {
+                if (throwable != null) {
+                    log.error("error while describing cluster", throwable);
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Successfully described cluster: " + nodes);
+                    }
+                    onClusterDescribeSuccess(nodes);
+                }
```

Comment:
This is still being executed on the Admin client's response handler thread. Ultimately the call to `volumeObserver.observeVolumeUsage(volumes);` also runs on the Admin client's response handler thread, so it ought not to block, but we don't actually know how long it might block for (because some of the factor policies invoke `Runnable#run` in `observeVolumeUsage`). Now, I know _in practice_ we're not expecting it to block, because the `listener.run()` is just adding something to a collection. 

TBH I think it's just better to avoid using `whenComplete()` _at all_ in this class, and just write get the Admin results by blocking on the thread using `get`. I think that would give you what you actually intend by the scheduled execution of this `run()`. It also simpler code.

Reply:
Resolved with 876b9a0b102c96ea8c946b3d471684472c109976